### ISSUE TITLE
Improve the "not exhaustive" diagnostic in the presence of a when clause.

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#45157]
After https://github.com/dotnet/roslyn/pull/44702 we produce a diagnostic that gives an example of a pattern that would cover a non-covered scenario in a switch expression that isn't exhaustive. However, we don't tell the user when that is because of the presence of a `when` clause. @alrz observed this [in code review](https://github.com/dotnet/roslyn/pull/44702#issuecomment-641706686):

> I wonder what example we give for:
> ```cs
> return unsigned switch 
> {
>   0 => true,
>   var x when x > 0 => false,
> };
> ```
> I would expect `_` since any other number is caught by the second arm. F# gives the example anyways but has a special diagnostic in presence of when. `For example, the value '1' may indicate a case not covered by the pattern(s). However, a pattern rule with a 'when' clause might successfully match this value.`

We should definitely enhance the diagnostic to include this information when appropriate.
